### PR TITLE
Entities/appoinments

### LINKED ISSUE #27: creating appointments
At the beginning of a new semester, a VLA member clicks somewhere (tbd) in the calendar, and a page to create a new appointment opens. This form has fields to enter 
* the title
* the lecture type it belongs to (optional) (#34)
* category (lecture/vacation/Abnahmetermin/...) (#28)
* start datetime, end datetime, recurrence week days, recurrence end date,
* peoples / lecturers: there may be multiple people (Prof, Assistenz) per appointment (#33)

After submitting the form, this appointment is visible in the main calendar view, colored according to its category.

---

Questions to be answered:

* [x] person: list of existing people or free-text form?

ToDo:

* [x] lectures -> #69 
* [x] appointment categories -> #70 
* [x] persons -> #71 
* [ ] lecture_persons
* [x] appointment_series -> #99 
* [x] appointments -> #99 
* [x] acceptance -> #99 
* [x] experiment_booking -> #99

### PR DESCRIPTION
**Pull request by JohaJung**
Introduces appointment_series, appointments and acceptance

One step further to close #27

---

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified without unreasonable effort
- [x] Is the code modular enough?
- [x] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [x] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [x] Interface Segregation: If interface implemented, is it fully implemented?
- [x] Dependency Injection: Are central object instantiation processes managed by a single context?
- [x] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [x] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?

### Dependencies (internal and external)

- [x] Were updates to documentation, configuration, or readme files made as required by this change?
- [x] Is the API.md up-to-date?
- [ ] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [ ] Does the code introduce any security vulnerabilities?
- [x] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is sensitive data like user data handled and stored securely?
- [ ] Is the right encryption used?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [x] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [ ] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [ ] Is the API/UI intuitive to use?
- [ ] Was the app tested on another device?

### Readability

- [x] Is the code easy to understand?
- [x] Which parts were confusing to you and why?
- [x] Can the readability of the code be improved by smaller methods/different names?
- [x] Is the code located in the right file/folder?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Are there redundant or outdated comments?
- [x] Could some comments convey the message better?
- [ ] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/acceptance/AcceptanceController.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     */
+    @PutMapping("/acceptances/{id}")
+    public ResponseEntity<?> updateAcceptance(@PathVariable Long id,
+                                              @RequestBody Acceptance acceptance) {
+        if (Objects.isNull(acceptance.getId())) {
+            acceptance.setId(id);
+        } else if (!acceptance.getId().equals(id)) {
+            LOGGER.warn("Received inconsistent IDs on acceptance modification." +
+                " ID from url: {} vs. ID from body data: {}.", id, acceptance.getId());
+            return ResponseEntity.badRequest().build();
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/acceptance/AcceptanceController.java]: Throw an exception

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/acceptance/AcceptanceController.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+     *
+     * @param id ID of the acceptance to delete.
+     * @return OK response with the deleted acceptance, Error response otherwise.
+     */
+    @DeleteMapping("/acceptances/{id}")
+    public ResponseEntity<?> deleteAcceptance(@PathVariable Long id) {
+        Optional<Acceptance> acceptanceOptional = acceptanceRepository.findById(id);
+        if (acceptanceOptional.isEmpty()) {
+            LOGGER.warn("Received acceptance deletion for non-existing acceptance with ID {}.", id);
+            return ResponseEntity.notFound().build();
```
[the remark this answers is not in the document]
Comment (MarcoLeipnitz) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/acceptance/AcceptanceController.java]: Or just replace with the orElse part of the Optional

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointment/AppointmentController.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    @PutMapping("/appointments/{id}")
+    public ResponseEntity<?> updateAppointment(@PathVariable Long id,
+                                               @RequestBody Appointment appointment) {
+        if (Objects.isNull(appointment.getId())) {
+            appointment.setId(id);
+        } else if (!appointment.getId().equals(id)) {
+            LOGGER.warn("Received inconsistent IDs on appointment modification." +
+                " ID from url: {} vs. ID from body data: {}.", id, appointment.getId());
+            return ResponseEntity.badRequest().build();
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointment/AppointmentController.java]: Throw an exception

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointment/AppointmentController.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     * @param appointment Dataset of the appointment to update.
+     *                   Must contain all keys, ID may be omitted.
+     * @return OK response with the updated appointment, Error response otherwise.
+     */
+    @PutMapping("/appointments/{id}")
+    public ResponseEntity<?> updateAppointment(@PathVariable Long id,
+                                               @RequestBody Appointment appointment) {
+        if (Objects.isNull(appointment.getId())) {
+            appointment.setId(id);
+        } else if (!appointment.getId().equals(id)) {
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointment/AppointmentController.java]: See Acceptance Controller

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointmentcategory/AppointmentCategoryController.java [hunk 5]
```diff
@@ -109,11 +96,8 @@ public ResponseEntity<?> deleteAppointmentCategory(@PathVariable Long id) {
...
         // TODO: use a better method here instead of debug message
-        SseController.notifyDebugTest("Appointment category deleted: " + id);
-        EntityModel<AppointmentCategory> appCatModel =
-            EntityModel.of(appointmentCategoryOptional.get(), linkTo(
-                methodOn(AppointmentCategoryController.class).deleteAppointmentCategory(
-                    id)).withSelfRel());
-        return ResponseEntity.ok(appCatModel);
+        SseController.notifyDebugTest(
+            "Appointment category deleted: " + appointmentCategoryOptional.get());
+        return ResponseEntity.ok(appointmentCategoryOptional.get());
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/appointmentcategory/AppointmentCategoryController.java]: This .get should not be done here anymore, take the code before and refactor it with a proper exception and unpacking.

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/lecture/LectureController.java [hunk 6]
```diff
@@ -99,9 +88,6 @@ public ResponseEntity<?> deleteLecture(@PathVariable Long id) {
         lectureRepository.deleteById(id);
         // TODO: use a better method here instead of debug message
         SseController.notifyDebugTest("Lecture deleted: " + lectureOptional.get());
-
-        EntityModel<Lecture> lectureModel = EntityModel.of(lectureOptional.get(),
-            linkTo(methodOn(LectureController.class).deleteLecture(id)).withSelfRel());
-        return ResponseEntity.ok(lectureModel);
+        return ResponseEntity.ok(lectureOptional.get());
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/entities/calendar/lecture/LectureController.java]: This would also benefit from earlier optional unpacking + error handling

**Review by MarcoLeipnitz — APPROVED** [review]
That looks good, just a few things I wanted to note

File: VLA-Backend/src/main/resources/db/migrations/V2__remaining_appointment_entities.sql [hunk 7]
```diff
@@ -0,0 +1,62 @@
+CREATE SEQUENCE IF NOT EXISTS acceptances_seq START WITH 1 INCREMENT BY 50;
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Backend/src/main/resources/db/migrations/V2__remaining_appointment_entities.sql]: @MarcoLeipnitz Is this enough?
https://github.com/TUDa-VLAOrga/VLAOrga/blob/main/docs/db-migrations.md#generating-a-migration

Will add a comment at the top of the migration file that it's auto-generated

**Comment by MarcoLeipnitz** [comment]
I have added a task that automatically generates the typescript interfaces, so that we can make changes without breaking the entire application at a later point.

Note that the Controllers and Respositories are carried over as well.
To avoid this, we would need to restructure the placement of repos and controller or just the entities.
There is no exclusion option without hardcoding the excluded classes, trust me, I tried..

**Review by CarolinRoell — APPROVED** [review]
